src/treemodel/selection.js

```diff
@@ -0,0 +1,261 @@
+/**
+ * @license Copyright (c) 2003-2015, CKSource - Frederico Knabben. All rights reserved.
+ * For licensing, see LICENSE.md.
+ */
+
+'use strict';
+
+CKEDITOR.define( [
+	'treemodel/liverange',
+	'treemodel/attributelist',
+	'emittermixin',
+	'utils',
+	'ckeditorerror'
+], ( LiveRange, AttributeList, EmitterMixin, utils, CKEditorError ) => {
+	/**
+	 * Represents a selection that is made on nodes in {@link treeModel.Document}. Selection instance is
+	 * created by {@link treeModel.Document}. In most scenarios you should not need to create an instance of Selection.
+	 *
+	 * @class treeModel.Selection
+	 */
+	class Selection {
+		/**
+		 * Creates an empty selection.
+		 *
+		 * @constructor
+		 */
+		constructor() {
+			/**
+			 * List of attributes set on current selection.
+			 *
+			 * @private
+			 * @property {treeModel.AttributeList} _attrs
+			 */
+			this._attrs = new AttributeList();
+
+			/**
+			 * Stores all ranges that are selected.
+			 *
+			 * @private
+			 * @property {Array.<LiveRange>}
+			 */
+			this._ranges = [];
+
+			/**
+			 * Specifies whether the last added range was added as a backward or forward range.
+			 *
+			 * @private
+			 * @property {Boolean}
+			 */
+			this._lastRangeBackward = false;
```

Comment:
If `anchor` and `focus` setting method changes we would have to deprecate `isBackward` and `isLastBackward` parameters. Then we would have to implement new way of setting them - they might be set automatically depending on new logic or set in explicit way. We might need to change getters to normal properties. These aren't big changes really. There are three things that we need to remember:
1) Right now our implementation is correct compared to Firefox. I have to admit that I haven't read W3C docs on how it should work, though. I trust that Firefox did it right. However, reading their docs, they don't have any explicit way to set the selection direction. You do it by setting range/collapsing selection and then by extending it. So they do it more behind the scenes.
2) It is rare that the user cares which way selection is set.
3) If anything becomes to complicated to change we still have the flexibility because this `Selection` class "lives" only in Tree Model space. It should have useful API so users can easily operate on it, but it will be "translated" into DOM Selection so we might do some hacking there. This is of course only last resort, if ever things get too complicated. Of course the primary goal is to be standards compilant.
BTW. If you would like me to make our `Selection` more like DOM one (API) let me know and I will change/add some methods.

As far as "backward" is concerned - I don't know, I think I just like when optional parameters default to `false`. For me it comes natural when I write a function. I can change it if you want :P.